Additional checks for HMC samplers

danielturek opened the pull request:
This PR adds additional error checks within the `NUTS` and `NUTS_classic` HMC samplers, for the case of assigning HMC samplers to invalid `target` nodes.  At present, this adds error traps for:

- HMC `target` being discrete
- HMC `target` having a truncated prior distribution
- HMC `calcNodes` having a truncated distribution
- HMC `calcNodes` having a `dinterval` or `dconstraint` distribution.

An open question relates to user-defined distributions which do not support AD calculations.  Presumably, assigning HMC to such a node should also be error trapped.  Is there an easy way to check if a user-defined distribution supports derivatives?

Do not merge, opening PR to relate this to issue #40.

perrydv commented:
Hi @danielturek,
The way to check whether a nimbleFunction has `buildDerivs` enabled is shown for example in nimble's `checkADsupportForDistribution`. It is
`isTRUE(environment(nf)$nfMethodRCobject[["buildDerivs"]])`
Now that you've asked, it strikes me that maybe nimble should have a generic function for this query.

danielturek commented:
@perrydv The `"buildDerivs"` element of the `nfMethodRCobject ` is in fact a list, so I modified your code to instead use `!isFALSE`.  Using `is.list()` I think would work fine too, 6 of one half dozen of another.

Relevant test also added.

danielturek commented:
Tests passing, merging this in.

perrydv commented:
@danielturek @paciorek I am following up here just after we've been working on a nimble release.

You were correct about `!isFALSE` instead of `isTRUE`.

`model$modelDef$checkADsupportForDistribution(dist)` should now (i.e. when nimble 1.2.0 is released) work. Note that `dist` should be obtained by `deparse(code)` or something like nimble's `safeDeparse(code)`, but not `as.character(code)` (which assumes `code` is simply a name).  (`deparse` is likely sufficent since `safeDeparse` is to handle very long expressions). The reason is that `code` could now be an expression like `object$dcustom`.

Also the `buildDerivs` element of a nimbleFunction or a nimbleFunction method that becomes registered as s distribution is now copied to a field called `buildDerivs` in the `distClass` object, i.e. in the result of `getDistributionInfo(dist)` (where again `dist` should be constructed by `deparse`). As you noted before, derivs are enabled if `buildDerivs` is either `TRUE` or a list.

Really `model$modelDef$checkADsupportForDistribution(dist)` is a bit misplaced (does not need to be in the `modelDef` object, so we might want to make it simply an R function.

I also made all built-in distributions have a `buildDerivs` field set to `TRUE` except those in an exception list, which is currently `dconstraint` and `dinterval`, IIRC.
